### PR TITLE
Extended keystore and parameters

### PR DESCRIPTION
**Pull request by israpps**
## Pull Request checklist

Note: these are not necessarily requirements

- [ ] I reformatted the code with clang-format
- [ ] I checked to make sure my submission worked
- [x] I am the author of submission or have permission from the original author
- [ ] Requires update of the PS2SDK or other dependencies
- [ ] Others (please specify below)

## Features:
- make the keystore file an ini file, allowing for sections, also, comments now will be supported (instead of triggering error)
- flags can now be passed to encrypt command, these are:
  + `--keys=KEY`: searches for the mg keys on the section provided, defaults to `default`
  + `--systemtype=TYPE`: manually sets the system type, supported values are `PS2`, `PSX` or a hex number .
  + `--kflags=FLAGS`: Custom KELF flags, supported values are `KELF`, `KIRX` or a hex number.
  + `--mgzone=VALUE`: custom MagicGate region zones, expects a hex number.
  + `--apptype=VALUE`: custom KELF application type, expects a hex number.

### new keystore format
```
[default]
MG_SIG_MASTER_KEY=...
MG_SIG_HASH_KEY=...
MG_KBIT_MASTER_KEY=...
MG_KBIT_IV=...
MG_KC_MASTER_KEY=...
MG_KC_IV=...
MG_ROOTSIG_MASTER_KEY=...
MG_ROOTSIG_HASH_KEY=...
MG_CONTENT_TABLE_IV=...
MG_CONTENT_IV=...
[arcade]
MG_SIG_MASTER_KEY=...
MG_SIG_HASH_KEY=...
MG_KBIT_MASTER_KEY=...
MG_KBIT_IV=...
MG_KC_MASTER_KEY=...
MG_KC_IV=...
MG_ROOTSIG_MASTER_KEY=...
MG_ROOTSIG_HASH_KEY=...
MG_CONTENT_TABLE_IV=...
MG_CONTENT_IV=...
[dev]
MG_SIG_MASTER_KEY=...
MG_SIG_HASH_KEY=...
MG_KBIT_MASTER_KEY=...
MG_KBIT_IV=...
MG_KC_MASTER_KEY=...
MG_KC_IV=...
MG_ROOTSIG_MASTER_KEY=...
MG_ROOTSIG_HASH_KEY=...
MG_CONTENT_TABLE_IV=...
MG_CONTENT_IV=...
```

In addition, `ARCADE_KBIT` & `ARCADE_KC` have been renamed to `OVERRIDE_KBIT` & `OVERRIDE_KC` for readability
And when these values are found and valid, program will explicitly inform they're being used by the following message:
```
-- kbit and Kc overridden by keystore!
```

## keystore search:
Now the keystore file from CWD will have priority over the one on $HOME.

## Decrypt command.
The decrypt command will also be capable of using the `--keys` flag. and it will now Identify namco arcade dongle boot.bin user header (not sure if we want to support it on the encrypt cmd??)

### CODE REVIEW

**Comment by israpps** [comment]
New look for decrypt command
IMG-20240403-WA0006

**Comment by israpps** [comment]
@AKuHAK btw, since balika is not quite active, what do you think about enabling discussions tab on this repo, creating a new one, and use that link to replace the report link of the program

File: src/common.h [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#define REDBOLD "\033[1;31m"
```
Comment (AKuHAK) [inline src/common.h:1]: Does this work in CMD? I think only powrshell supports that nice formatting. Actually, I was forced to remove it from hfl-dump due to many reports about this characters in cmd. Can you remove formatting for now, and add it later as separate PR?

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+# add the executable
+include_directories(src)
+find_package(OpenSSL REQUIRED)
+add_executable(kelftool 
+    src/kelf.cpp
+    src/kelftool.cpp
+    src/keystore.cpp
+)
+target_link_libraries(kelftool OpenSSL::SSL)
```
Comment (AKuHAK) [inline CMakeLists.txt:15]: So then adding this in GitHub action?

File: src/kelf.cpp [hunk 3]
```diff
@@ -97,142 +98,149 @@ int Kelf::LoadKelf(const std::string &filename)
 
     if (header.Flags & 1 || header.Flags & 0xf0000 || header.BitCount != 0) {
         // TODO: check more unknown bit flags
-        printf("This file is not supported yet and looked after.\n");
+        printf(REDBOLD "This file is not supported yet and looked after.\n");
```
Comment (AKuHAK) [inline src/kelf.cpp:101]: Its now super hard to check where you applied formatting, and where you actually changed the code.

**Review by AKuHAK — COMMENTED** [review]
As always commit history is dirty. It is ok when you change one thing, but here you changed a lot more:
- added shell formatting
- added new functionality
- added new library
- added cmake
- fixed some bugs

It can be in one pr, but then it requires clean commits, one commit per change. If it is not an option, then, please split this pr in pr series, so I can squash each one. Also, shell formatting is very questionable change, will require some testing.

**Review by AKuHAK — CHANGES_REQUESTED** [review]
.

**Comment by israpps** [comment]
> As always commit history is dirty. It is ok when you change one thing, but here you changed a lot more:
> - added shell formatting
> - added new functionality
> - added new library
> - added cmake
> - fixed some bugs
> 
> It can be in one pr, but then it requires clean commits, one commit per change. If it is not an option, then, please split this pr in pr series, so I can squash each one. Also, shell formatting is very questionable change, will require some testing.

I forgot the cmake there

It was for quicker compiling.

I can remove it. Or leave it, since it's not hooked to CI in any form

**Comment by israpps** [comment]
i'll squash the commits then

File: src/common.h (hunk 1 above)
Comment (israpps) [reply src/common.h:1]: > Does this work in CMD? I think only powrshell supports that nice formatting. Actually, I was forced to remove it from hfl-dump due to many reports about this characters in cmd. Can you remove formatting for now, and add it later as separate PR?

or disable on windows. althoug, WSL and windows terminal do support it

**Comment by israpps** [comment]
closed in favor of #9

**Pull request closed without merging**